File: android/app/src/main/java/com/cactus/remoteterminal/terminal/Row.kt
package com.cactus.remoteterminal.terminal

/**
 * One terminal row backed by parallel IntArrays (no per-cell objects).
 *
 * `codes[col]` is the code point shown in the cell (' ' for blank, 0 for the
 * right half of a wide glyph whose left half carries [TerminalEmulator.WIDE]).
 * Combining marks are rare, so they live in a lazily created map keyed by column.
 */
class Row(val cols: Int) {
    val codes = IntArray(cols) { BLANK }
    val fg = IntArray(cols) { TerminalEmulator.DEFAULT }
    val bg = IntArray(cols) { TerminalEmulator.DEFAULT }
    val flags = IntArray(cols)

    /** This row continues on the next one (soft wrap); used by reflow and selection. */
    var wrapped = false

    private var combiningMap: HashMap<Int, String>? = null

    fun combining(col: Int): String? = combiningMap?.get(col)

    fun setCombining(col: Int, marks: String?) {
        if (marks == null) { combiningMap?.remove(col); return }
        val m = combiningMap ?: HashMap<Int, String>().also { combiningMap = it }
        m[col] = marks
    }

    fun appendCombining(col: Int, mark: String) {
        val m = combiningMap ?: HashMap<Int, String>().also { combiningMap = it }
        val prev = m[col]
        m[col] = if (prev == null) mark else if (prev.length >= 16) prev else prev + mark
    }

    fun hasCombining(): Boolean = !combiningMap.isNullOrEmpty()

    fun set(col: Int, code: Int, fgc: Int, bgc: Int, fl: Int) {
        codes[col] = code; fg[col] = fgc; bg[col] = bgc; flags[col] = fl
        combiningMap?.remove(col)
    }

    /** Blank one cell with the given colours (background colour erase). */
    fun clear(col: Int, fgc: Int, bgc: Int) {
        codes[col] = BLANK; fg[col] = fgc; bg[col] = bgc; flags[col] = 0
        combiningMap?.remove(col)
    }

    /** Blank [from, to) with the given colours. */
    fun clearRange(from: Int, to: Int, fgc: Int, bgc: Int) {
        val a = from.coerceAtLeast(0)
        val b = to.coerceAtMost(cols)
        for (c in a until b) { codes[c] = BLANK; fg[c] = fgc; bg[c] = bgc; flags[c] = 0 }
        val m = combiningMap
        if (m != null && m.isNotEmpty()) for (c in a until b) m.remove(c)
    }

    fun fill(fgc: Int, bgc: Int) { clearRange(0, cols, fgc, bgc); wrapped = false }

    /** Copy cells [src, src+count) to [dst, dst+count) within this row (overlap-safe). */
    fun moveCells(src: Int, dst: Int, count: Int) {
        if (count <= 0 || src == dst) return
        System.arraycopy(codes, src, codes, dst, count)
        System.arraycopy(fg, src, fg, dst, count)
        System.arraycopy(bg, src, bg, dst, count)
        System.arraycopy(flags, src, flags, dst, count)
        val m = combiningMap
        if (m != null && m.isNotEmpty()) {
            val moved = HashMap<Int, String>()
            for ((k, v) in m) {
                if (k in src until src + count) moved[k - src + dst] = v
                else if (k !in dst until dst + count) moved[k] = v
            }
            combiningMap = moved
        }
    }

    fun copyFrom(other: Row) {
        val n = minOf(cols, other.cols)
        System.arraycopy(other.codes, 0, codes, 0, n)
        System.arraycopy(other.fg, 0, fg, 0, n)
        System.arraycopy(other.bg, 0, bg, 0, n)
        System.arraycopy(other.flags, 0, flags, 0, n)
        if (n < cols) clearRange(n, cols, TerminalEmulator.DEFAULT, TerminalEmulator.DEFAULT)
        wrapped = other.wrapped
        combiningMap = null
        val om = other.combiningMap
        if (om != null && om.isNotEmpty()) for ((k, v) in om) if (k < n) setCombining(k, v)
    }

    /** True when the cell is visually empty (blank glyph, default background, no attributes). */
    fun isBlankCell(col: Int): Boolean =
        codes[col] == BLANK && bg[col] == TerminalEmulator.DEFAULT && (flags[col] and (TerminalEmulator.UNDERLINE or TerminalEmulator.REVERSE or TerminalEmulator.STRIKE)) == 0 && combining(col) == null

    fun isBlank(): Boolean {
        for (c in 0 until cols) if (!isBlankCell(c)) return false
        return true
    }

    /** Index after the last non-blank cell (0 when the row is blank). */
    fun contentEnd(): Int {
        var end = cols
        while (end > 0 && isBlankCell(end - 1)) end--
        return end
    }

    /** Row text; continuation cells are skipped and combining marks are included. */
    fun text(trimEnd: Boolean = true): String = textRange(0, cols, trimEnd)

    fun textRange(from: Int, to: Int, trimEnd: Boolean): String {
        val sb = StringBuilder()
        appendText(sb, from, to, trimEnd)
        return sb.toString()
    }

    fun appendText(sb: StringBuilder, from: Int, to: Int, trimEnd: Boolean) {
        val a = from.coerceAtLeast(0)
        var b = to.coerceAtMost(cols)
        if (trimEnd) while (b > a && codes[b - 1] == BLANK && combining(b - 1) == null) b--
        for (c in a until b) {
            val code = codes[c]
            if (code == 0) continue // continuation of a wide glyph
            sb.appendCodePoint(code)
            combining(c)?.let { sb.append(it) }
        }
    }

    companion object {
        const val BLANK = ' '.code
    }
}

File: desktop/ui/js/terminal/row.js
/**
 * One terminal row backed by parallel typed arrays (no per-cell objects).
 *
 * `codes[col]` is the code point shown in the cell (' ' for blank, 0 for the
 * right half of a wide glyph whose left half carries `WIDE`). Combining marks
 * are rare, so they live in a lazily created map keyed by column.
 *
 * A port of `Row.kt`.
 */

import { DEFAULT, UNDERLINE, REVERSE, STRIKE } from './attrs.js';

export const BLANK = 0x20;

export class Row {
  constructor(cols) {
    this.cols = cols;
    this.codes = new Int32Array(cols).fill(BLANK);
    this.fg = new Int32Array(cols).fill(DEFAULT);
    this.bg = new Int32Array(cols).fill(DEFAULT);
    this.flags = new Int32Array(cols);
    /** This row continues on the next one (soft wrap); used by reflow and selection. */
    this.wrapped = false;
    this.combiningMap = null;
  }

  combining(col) {
    return this.combiningMap ? this.combiningMap.get(col) ?? null : null;
  }

  setCombining(col, marks) {
    if (marks == null) {
      if (this.combiningMap) this.combiningMap.delete(col);
      return;
    }
    if (!this.combiningMap) this.combiningMap = new Map();
    this.combiningMap.set(col, marks);
  }

  appendCombining(col, mark) {
    if (!this.combiningMap) this.combiningMap = new Map();
    const prev = this.combiningMap.get(col);
    this.combiningMap.set(col, prev == null ? mark : prev.length >= 16 ? prev : prev + mark);
  }

  hasCombining() {
    return !!this.combiningMap && this.combiningMap.size > 0;
  }

  set(col, code, fgc, bgc, fl) {
    this.codes[col] = code;
    this.fg[col] = fgc;
    this.bg[col] = bgc;
    this.flags[col] = fl;
    if (this.combiningMap) this.combiningMap.delete(col);
  }

  /** Blank one cell with the given colours (background colour erase). */
  clear(col, fgc, bgc) {
    this.codes[col] = BLANK;
    this.fg[col] = fgc;
    this.bg[col] = bgc;
    this.flags[col] = 0;
    if (this.combiningMap) this.combiningMap.delete(col);
  }

  /** Blank [from, to) with the given colours. */
  clearRange(from, to, fgc, bgc) {
    const a = Math.max(0, from);
    const b = Math.min(this.cols, to);
    for (let c = a; c < b; c++) {
      this.codes[c] = BLANK;
      this.fg[c] = fgc;
      this.bg[c] = bgc;
      this.flags[c] = 0;
    }
    if (this.combiningMap && this.combiningMap.size > 0) {
      for (let c = a; c < b; c++) this.combiningMap.delete(c);
    }
  }

  fill(fgc, bgc) {
    this.clearRange(0, this.cols, fgc, bgc);
    this.wrapped = false;
  }

  /** Copy cells [src, src+count) to [dst, dst+count) within this row (overlap-safe). */
  moveCells(src, dst, count) {
    if (count <= 0 || src === dst) return;
    this.codes.copyWithin(dst, src, src + count);
    this.fg.copyWithin(dst, src, src + count);
    this.bg.copyWithin(dst, src, src + count);
    this.flags.copyWithin(dst, src, src + count);
    const m = this.combiningMap;
    if (m && m.size > 0) {
      const moved = new Map();
      for (const [k, v] of m) {
        if (k >= src && k < src + count) moved.set(k - src + dst, v);
        else if (!(k >= dst && k < dst + count)) moved.set(k, v);
      }
      this.combiningMap = moved;
    }
  }

  copyFrom(other) {
    const n = Math.min(this.cols, other.cols);
    this.codes.set(other.codes.subarray(0, n), 0);
    this.fg.set(other.fg.subarray(0, n), 0);
    this.bg.set(other.bg.subarray(0, n), 0);
    this.flags.set(other.flags.subarray(0, n), 0);
    if (n < this.cols) this.clearRange(n, this.cols, DEFAULT, DEFAULT);
    this.wrapped = other.wrapped;
    this.combiningMap = null;
    const om = other.combiningMap;
    if (om && om.size > 0) for (const [k, v] of om) if (k < n) this.setCombining(k, v);
  }

  /** True when the cell is visually empty (blank glyph, default background, no attributes). */
  isBlankCell(col) {
    return (
      this.codes[col] === BLANK &&
      this.bg[col] === DEFAULT &&
      (this.flags[col] & (UNDERLINE | REVERSE | STRIKE)) === 0 &&
      this.combining(col) == null
    );
  }

  isBlank() {
    for (let c = 0; c < this.cols; c++) if (!this.isBlankCell(c)) return false;
    return true;
  }

  /** Index after the last non-blank cell (0 when the row is blank). */
  contentEnd() {
    let end = this.cols;
    while (end > 0 && this.isBlankCell(end - 1)) end--;
    return end;
  }

  /** Row text; continuation cells are skipped and combining marks are included. */
  text(trimEnd = true) {
    return this.textRange(0, this.cols, trimEnd);
  }

  textRange(from, to, trimEnd) {
    const out = [];
    this.appendText(out, from, to, trimEnd);
    return out.join('');
  }

  /** Append the row's text into [out]; `out` is an array of string pieces. */
  appendText(out, from, to, trimEnd) {
    const a = Math.max(0, from);
    let b = Math.min(this.cols, to);
    if (trimEnd) {
      while (b > a && this.codes[b - 1] === BLANK && this.combining(b - 1) == null) b--;
    }
    for (let c = a; c < b; c++) {
      const code = this.codes[c];
      if (code === 0) continue; // continuation of a wide glyph
      out.push(String.fromCodePoint(code));
      const marks = this.combining(c);
      if (marks) out.push(marks);
    }
  }
}
